### PR TITLE
module 3 - severless api

### PR DESCRIPTION
**Pull request by yulei-chen**
## Completion

✔️  1 - Product Service Serverless config contains configuration for 2 lambda functions, ~~API is not working at all, but~~ YAML configuration is correct
✔️ 2 - The getProductsList OR getProductsById lambda function returns a correct response (POINT1)
✔️ 3 - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2)
✔️ 4 - Your own Frontend application is integrated with Product Service (/products API) and products from Product Service are represented on Frontend. AND POINT1 and POINT2 are done.

Additional (optional) tasks:
✔️ +1 - Async/await is used in lambda functions
✔️ +1 - ES6 modules are used for Product Service implementation
💭 +1 - Webpack is configured for Product Service.  
**ESBuild is already integrated in the serverless template. You can see the esbuild config as following in `serverless.ts`.**
image

✔️ +1 (All languages) - SWAGGER documentation is created for Product Service. 
**Click [here](https://w0dhw8ytcl.execute-api.ap-northeast-2.amazonaws.com/dev/api) to see the document.**

✔️ +1 (All languages) - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered).
**I wrote tests for function `getProductsById` using `Jest`. You can `cd product-service` and run `npm test` to see the following results.**
image

✔️ +1 (All languages) - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase.
✔️ +1 (All languages) - Main error scenarios are handled by API ("Product not found" error).
**Click [here](https://w0dhw8ytcl.execute-api.ap-northeast-2.amazonaws.com/dev/products/nonexisitent-id) to get the error.**

## Product Service API

getProductsList: GET - https://w0dhw8ytcl.execute-api.ap-northeast-2.amazonaws.com/dev/products
getProductsById: GET - https://w0dhw8ytcl.execute-api.ap-northeast-2.amazonaws.com/dev/products/P0001

## FE PR

https://github.com/yuleicul/shop-react-redux-cloudfront/pull/2

### CODE REVIEW

**Comment by Mrinalaot** [comment]
1 - Product Service Serverless config contains configuration for 2 lambda functions, API is not working at all, but YAML configuration is correct - DONE
2 - The getProductsList OR getProductsById lambda function returns a correct response (POINT1) - DONE
3 - The getProductsById AND getProductsList lambda functions return a correct response code (POINT2) - DONE
4 - Your own Frontend application is integrated with Product Service (/products API) and products from Product Service are represented on Frontend. AND POINT1 and POINT2 are done. - DONE

+1 - Async/await is used in lambda functions - DONE
+1 - ES6 modules are used for Product Service implementation - DONE
+1 - Webpack is configured for Product Service - NOT DONE
+1 (All languages) - SWAGGER documentation is created for Product Service - DONE
+1 (All languages) - Lambda handlers are covered by basic UNIT tests (NO infrastructure logic is needed to be covered) - DONE
+1 (All languages) - Lambda handlers (getProductsList, getProductsById) code is written not in 1 single module (file) and separated in codebase. - DONE
+1 (All languages) - Main error scenarios are handled by API ("Product not found" error). - DONE

File: product-service/src/functions/getProductsById/handler.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import { middyfy } from "@libs/lambda";
+import { NotFoundException } from "src/classes/NotFoundException";
+import { PRODUCT_LIST } from "src/mock";
+
+export const findProductById = (productId: string) => {
+  return PRODUCT_LIST.find(item => productId === item.id)
+}
+
+const getProductsById: ValidatedEventAPIGatewayProxyEvent<unknown> = async (event) => {
+  const productId = event.pathParameters?.productId!
```
Comment (xdMus) [inline product-service/src/functions/getProductsById/handler.ts:11]: This is bad practice to use `!`. You should check is it exists and if it's not - return an error.

File: product-service/src/functions/getProductsById/handler.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  const product = findProductById(productId)
+
+  if (product) return formatJSONResponse(product)
+  else return new NotFoundException('Product not found')
```
Comment (xdMus) [inline product-service/src/functions/getProductsById/handler.ts:16]: The main error scenario isn't covered like some unexpected error - `500` status code. Please use `try/catch` for this cases

File: product-service/src/functions/getProductsList/handler.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { formatJSONResponse, ValidatedEventAPIGatewayProxyEvent } from "@libs/apiGateway";
+import { middyfy } from "@libs/lambda";
+import { PRODUCT_LIST } from "src/mock";
+
+const getProductsList: ValidatedEventAPIGatewayProxyEvent<unknown> = async () => {
+  return formatJSONResponse(PRODUCT_LIST)
+}
```
Comment (xdMus) [inline product-service/src/functions/getProductsList/handler.ts:7]: This function isn't covered with error handlers at all. Please fix it

File: product-service/src/functions/getProductsById/handler.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { formatJSONResponse, ValidatedEventAPIGatewayProxyEvent } from "@libs/apiGateway";
+import { middyfy } from "@libs/lambda";
+import { NotFoundException } from "src/classes/NotFoundException";
+import { PRODUCT_LIST } from "src/mock";
+
+export const findProductById = (productId: string) => {
+  return PRODUCT_LIST.find(item => productId === item.id)
+}
```
Comment (xdMus) [inline product-service/src/functions/getProductsById/handler.ts:8]: I don't see any `async/await` cases for your lambdas. This function could be asynchronous (and will be in the next tasks) and work as a service. This is a way to use `await` for your lambda

File: product-service/src/functions/getProductsList/handler.ts (hunk 3 above)
Comment (xdMus) [inline product-service/src/functions/getProductsList/handler.ts:7]: Please add unit tests for this lambda

File: product-service/src/libs/apiGateway.ts [hunk 5]
```diff
@@ -4,9 +4,15 @@ import type { FromSchema } from "json-schema-to-ts";
 type ValidatedAPIGatewayProxyEvent<S> = Omit<APIGatewayProxyEvent, 'body'> & { body: FromSchema<S> }
 export type ValidatedEventAPIGatewayProxyEvent<S> = Handler<ValidatedAPIGatewayProxyEvent<S>, APIGatewayProxyResult>
 
-export const formatJSONResponse = (response: Record<string, unknown>) => {
+export const formatJSONResponse = (response: Record<string, unknown> | Array<unknown>) => {
```
Comment (xdMus) [inline product-service/src/libs/apiGateway.ts:7]: You can use just `unknown` type for your response

File: product-service/src/functions/getProductsById/handler.ts (hunk 1 above)
Comment (yulei-chen) [reply product-service/src/functions/getProductsById/handler.ts:11]: Thank you sp much!! I'll fix all these issues in next task. @xdMus

**Pull request merged**